# EDA - Prepare Data for Model

## Load Librabries

In [1]:
import pandas as pd
import Functions_Books as funct
import nltk as nltk

from nltk.corpus import stopwords #Natural Language Toolkit
from nltk.tokenize import RegexpTokenizer
#nltk.download('RegexpTokenizer')

#import spacy.cli
#spacy.cli.download("fr_core_news_sm")
#FrenchNLP = spacy.load("fr_core_news_sm")

## Read cleaned data into dataframe

In [2]:
data_cleaned=pd.read_csv('DataFinal\DataCleanedFinal.csv',index_col=0)
data_preproc=data_cleaned.copy()

In [3]:
data_preproc.head()

title                 series  \
0                  To Kill a Mockingbird  To Kill a Mockingbird   
1  Harry Potter and the Sorcerer's Stone        Harry Potter #1   
2              The Diary of a Young Girl                   none   
3                            Animal Farm                   none   
4                      The Little Prince                   none   

                                             authors           isbn  \
0                                         Harper Lee     1439550417   
1                                       J.K. Rowling              0   
2  Anne Frank, Eleanor Roosevelt, Barbara Mooyaar...      330107372   
3       George Orwell, Russell Baker, C.M. Woodhouse  9780451526342   
4           Antoine de Saint-Exupéry, Richard Howard     3125971403   

                                              genres  \
0  Classics, Fiction, Historical, Historical Fict...   
1  Fantasy, Fiction, Young Adult, Magic, Children...   
2  Nonfiction, Classics, History, Biography, Auto...   
3  Classics, Fiction, Science Fiction, Dystopia, ...   
4  Classics, Fiction, Fantasy, Childrens, Cultura...   

                                         description  pages  year language  \
0  The unforgettable novel of a childhood in a sl...    324  1960  English   
1  Harry Potter's life is miserable. His parents ...    309  1997  English   
2  Discovered in the attic in which she spent the...    283  1947  English   
3  Librarian's note: There is an Alternate Cover ...    141  1945  English   
4  A pilot stranded in the desert awakes one morn...     96  1943  English   

                                         cover_image  total_number_ratings  \
0  https://i.gr-assets.com/images/S/compressed.ph...               5011381   
1  https://i.gr-assets.com/images/S/compressed.ph...               8003482   
2  https://i.gr-assets.com/images/S/compressed.ph...               3088199   
3  https://i.gr-assets.com/images/S/compressed.ph...               3095756   
4  https://i.gr-assets.com/images/S/compressed.ph...               1609561   

   average_rating  
0            4.27  
1            4.47  
2            4.16  
3            3.97  
4            4.31

In [4]:
data_preproc.shape

(8431, 12)

## Text Preprocessing - Tokenizing with NLTK

Info found here: https://towardsdatascience.com/5-simple-ways-to-tokenize-text-in-python-92c6804edfc4


### Add column Describtion_tokenized and Title_tokenized

In [5]:
selected_columns = data_preproc[["title","description","series"]]
newdf=selected_columns.copy()
newdf = newdf.rename(columns={'title': 'title_tokenized', 'description': 'desccription_tokenized', 'series':'series_tokenized'})
frames=[data_preproc,newdf]
DataForPreprocessing = pd.concat(frames,axis=1)
DataForPreprocessing.head()

title                 series  \
0                  To Kill a Mockingbird  To Kill a Mockingbird   
1  Harry Potter and the Sorcerer's Stone        Harry Potter #1   
2              The Diary of a Young Girl                   none   
3                            Animal Farm                   none   
4                      The Little Prince                   none   

                                             authors           isbn  \
0                                         Harper Lee     1439550417   
1                                       J.K. Rowling              0   
2  Anne Frank, Eleanor Roosevelt, Barbara Mooyaar...      330107372   
3       George Orwell, Russell Baker, C.M. Woodhouse  9780451526342   
4           Antoine de Saint-Exupéry, Richard Howard     3125971403   

                                              genres  \
0  Classics, Fiction, Historical, Historical Fict...   
1  Fantasy, Fiction, Young Adult, Magic, Children...   
2  Nonfiction, Classics, History, Biography, Auto...   
3  Classics, Fiction, Science Fiction, Dystopia, ...   
4  Classics, Fiction, Fantasy, Childrens, Cultura...   

                                         description  pages  year language  \
0  The unforgettable novel of a childhood in a sl...    324  1960  English   
1  Harry Potter's life is miserable. His parents ...    309  1997  English   
2  Discovered in the attic in which she spent the...    283  1947  English   
3  Librarian's note: There is an Alternate Cover ...    141  1945  English   
4  A pilot stranded in the desert awakes one morn...     96  1943  English   

                                         cover_image  total_number_ratings  \
0  https://i.gr-assets.com/images/S/compressed.ph...               5011381   
1  https://i.gr-assets.com/images/S/compressed.ph...               8003482   
2  https://i.gr-assets.com/images/S/compressed.ph...               3088199   
3  https://i.gr-assets.com/images/S/compressed.ph...               3095756   
4  https://i.gr-assets.com/images/S/compressed.ph...               1609561   

   average_rating                        title_tokenized  \
0            4.27                  To Kill a Mockingbird   
1            4.47  Harry Potter and the Sorcerer's Stone   
2            4.16              The Diary of a Young Girl   
3            3.97                            Animal Farm   
4            4.31                      The Little Prince   

                              desccription_tokenized       series_tokenized  
0  The unforgettable novel of a childhood in a sl...  To Kill a Mockingbird  
1  Harry Potter's life is miserable. His parents ...        Harry Potter #1  
2  Discovered in the attic in which she spent the...                   none  
3  Librarian's note: There is an Alternate Cover ...                   none  
4  A pilot stranded in the desert awakes one morn...                   none

### Explore Book Languages

We have books in different languages. Due to time reasons we will only process the English books with NLTK.
All other languages could be processed with spaCy at a later point in time.<br>


In [6]:
DataForPreprocessing['language'].value_counts()

English               7606
none                   542
Arabic                  55
French                  37
German                  31
Spanish                 29
Persian                 17
Portuguese              16
Dutch                   14
Polish                  12
Italian                  7
Romanian                 7
Turkish                  7
Indonesian               6
Multiple languages       5
Swedish                  4
Greek                    4
Tamil                    3
Malayalam                2
Finnish                  2
Hindi                    2
Filipino, Pilipino       2
Ukrainian                2
Russian                  2
Urdu                     2
Norwegian                1
Telugu                   1
Catalan, Valencian       1
Hungarian                1
Tagalog                  1
Serbian                  1
Bulgarian                1
Scots                    1
Kannada                  1
Japanese                 1
Danish                   1
Croatian                 1
M

###  Deal with different languages by preprocessing title

#### Explore books with language 'None'

In [7]:
NoneLangBooks=pd.DataFrame(DataForPreprocessing[DataForPreprocessing['language']=='none'])
NoneLangBooks.shape

(542, 15)

In [8]:
#Return lower cases
funct.return_Serieslowercase(NoneLangBooks,'title_tokenized')

In [9]:
#Remove stop words
NoneLangBooks['title_tokenized'] = NoneLangBooks.title_tokenized.apply(func=funct.remove_stopwords)

In [10]:
#Remove non-Ascii Characters
NoneLangBooks['title_tokenized'] = NoneLangBooks['title_tokenized'].apply(func=funct.remove_nonAscii)

In [11]:
#Remove punctuation
NoneLangBooks['title_tokenized'] = NoneLangBooks['title_tokenized'].apply(func=funct.remove_punctuation)
NoneLangBooks[NoneLangBooks['title_tokenized']=='']

title series                                   authors  \
987             كلاب الراعي   none                             أشرف العشماوي   
1158           سيدة الزمالك   none                             أشرف العشماوي   
3633           سرقات مشروعة   none                             أشرف العشماوي   
5103               سال بلوا   none                               عباس معروفی   
5157  تراژدی اسبی که ایستاد   none                            روح اله سهرابی   
5970              All of Me   none                            Kevine Walcott   
7131                मधुशाला   none  Harivansh Rai Bachchan, हरिवंश राय बच्चन   
8393              أرض الغجر   none                             أحمد أبو هيبة   

               isbn                                             genres  \
987               0             Novels, Historical, Historical Fiction   
1158     9777951620  Novels, Fiction, Historical, Historical Fictio...   
3633              0        History, Nonfiction, Northern Africa, Egypt   
5103     9463113744  Novels, Fiction, Cultural, Iran, Literature, R...   
5157              0                                               none   
5970  9781095642580  Contemporary, Novels, Fiction, Epic, Drama, Su...   
7131  9780140120097  Poetry, Classics, Cultural, India, Asian Liter...   
8393              0                                               none   

                                            description  pages  year language  \
987   .. هذه الرواية تأسرك منذ فصولها الأولى ، وقد أ...    355  2015     none   
1158  ليلة الحادث تسلل خمسة أشخاص لفيلا "سولومون شيك...    375  2018     none   
3633  يستعرض هذا الكتاب صفحات مجهولة من تاريخ سرقة و...    312  2011     none   
5103  می‌دانی اولین بوسهٔ جهان چه‌طور کشف شد؟دست‌هاش...    342  1992     none   
5157  در بخشی از این رمان می‌خوانیم:با این درخواست گ...    200  2013     none   
5970  Kevine Walcott's 'All of Me' is a holistic, bo...     38  2019     none   
7131  Dr Harivanshrai Bachchan is India's leading Hi...    152  1935     none   
8393                                               None    239  2014     none   

                                            cover_image  total_number_ratings  \
987   https://i.gr-assets.com/images/S/compressed.ph...                  7279   
1158  https://i.gr-assets.com/images/S/compressed.ph...                 12414   
3633  https://i.gr-assets.com/images/S/compressed.ph...                   988   
5103  https://i.gr-assets.com/images/S/compressed.ph...                  5515   
5157  https://i.gr-assets.com/images/S/compressed.ph...                   280   
5970                                                NaN                   761   
7131  https://i.gr-assets.com/images/S/compressed.ph...                  3288   
8393  https://i.gr-assets.com/images/S/compressed.ph...                   180   

      average_rating title_tokenized  \
987             4.16                   
1158            4.25                   
3633            4.26                   
5103            3.94                   
5157            1.91                   
5970            4.42                   
7131            4.45                   
8393            3.86                   

                                 desccription_tokenized series_tokenized  
987   .. هذه الرواية تأسرك منذ فصولها الأولى ، وقد أ...             none  
1158  ليلة الحادث تسلل خمسة أشخاص لفيلا "سولومون شيك...             none  
3633  يستعرض هذا الكتاب صفحات مجهولة من تاريخ سرقة و...             none  
5103  می‌دانی اولین بوسهٔ جهان چه‌طور کشف شد؟دست‌هاش...             none  
5157  در بخشی از این رمان می‌خوانیم:با این درخواست گ...             none  
5970  Kevine Walcott's 'All of Me' is a holistic, bo...             none  
7131  Dr Harivanshrai Bachchan is India's leading Hi...             none  
8393                                               None             none

 #### Remove non-English books

The result above shows all non-English books after tokenizing.

In [12]:
NoneLangBooks=NoneLangBooks[NoneLangBooks['title_tokenized']!='']
NoneLangBooks

title     series  \
283                                Mistress Suffragette       none   
350                          Stones (One True Child #5)       none   
355                                    Mission: Subhero       none   
356   A Summer of Good-Byes (Blue Triangle Press Boo...       none   
357                                   The Scopas Factor       none   
...                                                 ...        ...   
8388                   Aj's Ireland: A Christmas Comedy  A.J. #0.5   
8389                                             Shift!       none   
8397  Talent for Humanity: Stories of Creativity, Co...       none   
8399               Ola: The Tale of a Young Moon Maiden       none   
8424                                        Dear Lover,       none   

                  authors           isbn  \
283        Diana   Forbes     1946409073   
350             L.C. Conn              0   
355    Linda    Armstrong              0   
356        Fred Misurella              0   
357    Vincent Panettiere  9781543947717   
...                   ...            ...   
8388           Renee Riva  9780692329047   
8389  Bill C.  Castengera  9781500887032   
8397      Patrick Gaffney  9781626341814   
8399           D.C. Nkado     9781312035   
8424  Lori Jenessa Nelson  9781311448422   

                                                 genres  \
283   Fiction, Novels, Drama, Contemporary, Historic...   
350     Book Club, Novels, Fiction, Drama, Contemporary   
355   Contemporary, Novels, Fiction, Drama, Book Clu...   
356   Fiction, Novels, Adult, Contemporary, Adventur...   
357   Fiction, Contemporary, Novels, Drama, Book Clu...   
...                                                 ...   
8388                           Humor, Family, Adventure   
8389                                               none   
8397                              Self Help, Nonfiction   
8399                                               none   
8424                                    Poetry, Romance   

                                            description  pages  year language  \
283   A young woman without prospects at a ball in G...    333  2017     none   
350   Deep within the New Zealand bush, lies a plate...    327  2019     none   
355   With just a few lines of code, one average man...    225  2017     none   
356   PASSION IN PROVENCE:Ben and Lee Alto follow Va...      0  2017     none   
357   A Hmong "story cloth," a Revolutionary War bat...    326  2018     none   
...                                                 ...    ...   ...      ...   
8388  In 1967, nine-year-old AJ wins a trip to Irela...    174  2014     none   
8389  Humankind is evolving. Technology has advanced...    336  2014     none   
8397  We are all born with the power to imagine what...    208  2015     none   
8399  Set in the distant land of Odu, the witches of...      0  2014     none   
8424  Dear Lover –poems of love, loss, and disappoin...      0  2014     none   

                                            cover_image  total_number_ratings  \
283   https://i.gr-assets.com/images/S/compressed.ph...                 12973   
350   https://i.gr-assets.com/images/S/compressed.ph...                   569   
355   https://i.gr-assets.com/images/S/compressed.ph...                  2417   
356   https://i.gr-assets.com/images/S/compressed.ph...                  1713   
357   https://i.gr-assets.com/images/S/compressed.ph...                  1080   
...                                                 ...                   ...   
8388  https://i.gr-assets.com/images/S/compressed.ph...                    23   
8389  https://i.gr-assets.com/images/S/compressed.ph...                     5   
8397  https://i.gr-assets.com/images/S/compressed.ph...                    24   
8399  https://i.gr-assets.com/images/S/compressed.ph...                     2   
8424  https://i.gr-assets.com/images/S/compressed.ph...                    39   

      average_rating     

#### Set language to English for English books

In [13]:
NoneLangBooks['language']='English'
NoneLangBooks.head()

title series  \
283                               Mistress Suffragette   none   
350                         Stones (One True Child #5)   none   
355                                   Mission: Subhero   none   
356  A Summer of Good-Byes (Blue Triangle Press Boo...   none   
357                                  The Scopas Factor   none   

                authors           isbn  \
283      Diana   Forbes     1946409073   
350           L.C. Conn              0   
355  Linda    Armstrong              0   
356      Fred Misurella              0   
357  Vincent Panettiere  9781543947717   

                                                genres  \
283  Fiction, Novels, Drama, Contemporary, Historic...   
350    Book Club, Novels, Fiction, Drama, Contemporary   
355  Contemporary, Novels, Fiction, Drama, Book Clu...   
356  Fiction, Novels, Adult, Contemporary, Adventur...   
357  Fiction, Contemporary, Novels, Drama, Book Clu...   

                                           description  pages  year language  \
283  A young woman without prospects at a ball in G...    333  2017  English   
350  Deep within the New Zealand bush, lies a plate...    327  2019  English   
355  With just a few lines of code, one average man...    225  2017  English   
356  PASSION IN PROVENCE:Ben and Lee Alto follow Va...      0  2017  English   
357  A Hmong "story cloth," a Revolutionary War bat...    326  2018  English   

                                           cover_image  total_number_ratings  \
283  https://i.gr-assets.com/images/S/compressed.ph...                 12973   
350  https://i.gr-assets.com/images/S/compressed.ph...                   569   
355  https://i.gr-assets.com/images/S/compressed.ph...                  2417   
356  https://i.gr-assets.com/images/S/compressed.ph...                  1713   
357  https://i.gr-assets.com/images/S/compressed.ph...                  1080   

     average_rating                              title_tokenized  \
283            4.32                         mistress suffragette   
350            4.31                      stones one true child 5   
355            4.35                              mission subhero   
356            4.36  summer good byes blue triangle press book 2   
357            4.31                                scopas factor   

                                desccription_tokenized series_tokenized  
283  A young woman without prospects at a ball in G...             none  
350  Deep within the New Zealand bush, lies a plate...             none  
355  With just a few lines of code, one average man...             none  
356  PASSION IN PROVENCE:Ben and Lee Alto follow Va...             none  
357  A Hmong "story cloth," a Revolutionary War bat...             none

#### English books 
Concat the 2 dataframes with English books and tokenize.

In [14]:
EnglishBooks1=pd.DataFrame(DataForPreprocessing[DataForPreprocessing['language']=='English'])

In [15]:
frames=[NoneLangBooks,EnglishBooks1]
EnglishBooks = pd.concat(frames,axis=0)
EnglishBooks['language'].value_counts()

English    8140
Name: language, dtype: int64

In [16]:
EnglishBooks

title  \
283                                Mistress Suffragette   
350                          Stones (One True Child #5)   
355                                    Mission: Subhero   
356   A Summer of Good-Byes (Blue Triangle Press Boo...   
357                                   The Scopas Factor   
...                                                 ...   
8426                                   The Stove-Junker   
8427                                     The Treemakers   
8428        100 Years: A Journey to End a Vicious Cycle   
8429    Thrive: 30 Inspirational Rags-to-Riches Stories   
8430                                  The Smoke at Dawn   

                                        series               authors  \
283                                       none        Diana   Forbes   
350                                       none             L.C. Conn   
355                                       none    Linda    Armstrong   
356                                       none        Fred Misurella   
357                                       none    Vincent Panettiere   
...                                        ...                   ...   
8426                                      none            S.K. Kalsi   
8427                 The Treemakers Trilogy #1  Christina L. Rozelle   
8428                                      none       Mark L. Baynard   
8429                                      none         Jason Navallo   
8430  Civil War: 1861-1865, Western Theater #3           Jeff Shaara   

               isbn                                             genres  \
283      1946409073  Fiction, Novels, Drama, Contemporary, Historic...   
350               0    Book Club, Novels, Fiction, Drama, Contemporary   
355               0  Contemporary, Novels, Fiction, Drama, Book Clu...   
356               0  Fiction, Novels, Adult, Contemporary, Adventur...   
357   9781543947717  Fiction, Contemporary, Novels, Drama, Book Clu...   
...             ...                                                ...   
8426  9780990779063                                               none   
8427     1508840466  Science Fiction, Dystopia, Fantasy, Young Adul...   
8428  9780986138003                                               none   
8429     1511503882  Nonfiction, Self Help, Inspirational, Business...   
8430  9780345527417  Historical, Historical Fiction, Military Histo...   

                                            description  pages  year language  \
283   A young woman without prospects at a ball in G...    333  2017  English   
350   Deep within the New Zealand bush, lies a plate...    327  2019  English   
355   With just a few lines of code, one average man...    225  2017  English   
356   PASSION IN PROVENCE:Ben and Lee Alto follow Va...      0  2017  English   
357   A Hmong "story cloth," a Revolutionary War bat...    326  2018  English   
...                                                 ...    ...   ...      ...   
8426  In the winter of 2012, 79-year-old Somerset Ga...    338  2015  English   
8427  This is a previously-published alternate cover...    288  2014  English   
8428  100 Years, A Journey to End a Vicious Cycle is...    290  2015  English   
8429  Thrive: 30 Inspirational Rags-to-Riches Storie...    100  2015  English   
8430  NEW YORK TIMESBESTSELLERJeff Shaara returns to...    495  2014  English   

                                            cover_image  total_number_ratings  \
283   https://i.gr-assets.com/images/S/compressed.ph...                 12973   
350   https://i.gr-assets.com/images/S/compressed.ph...                   569   
355   https://i.gr-assets.com/images/S/compressed.ph...                  2417   
356   https://i.gr-assets.com/images/S/compressed.ph...                  1713   
357   https://i.gr-assets.com/images/S/compressed.ph...                  1080   
...                                                 ...                   ...   
8426  https://i.gr-assets.com/images/S/compress

In [17]:
#Return lower cases
funct.return_Serieslowercase(EnglishBooks,'title_tokenized')

In [18]:
#Remove stop words
EnglishBooks['title_tokenized'] = EnglishBooks.title_tokenized.apply(func=funct.remove_stopwords)

In [19]:
#Remove non-Ascii Characters
EnglishBooks['title_tokenized'] = EnglishBooks['title_tokenized'].apply(func=funct.remove_nonAscii)

In [20]:
#Remove punctuation
EnglishBooks['title_tokenized'] = EnglishBooks['title_tokenized'].apply(func=funct.remove_punctuation)

In [21]:
EnglishBooks[EnglishBooks['title_tokenized']=='']

title            series  \
1765           Why Not Me?              none   
285                     It              none   
905                    Out              none   
1190  Before We Were Yours              none   
1243             After You  Me Before You #2   
1322         Me Before You  Me Before You #1   
2285        More Than This              none   
2979   جاناتان، مرغ دریایی              none   
3391      It Had to Be You  Chicago Stars #1   
4227                    We              none   
4345      What Is the What              none   
4567           Being There              none   
4902                  Once           Once #1   
5836                   You            You #1   
6497                 After              none   

                               authors           isbn  \
1765                      Mindy Kaling  9780804138161   
285                       Stephen King  9780450411434   
905   Natsuo Kirino, Stephen    Snyder  9781400078370   
1190                      Lisa Wingate  9780425284681   
1243                        Jojo Moyes  9780143108863   
1322                        Jojo Moyes  9780143124542   
2285                      Patrick Ness  9781406350487   
2979         Richard Bach, فاطمه محمدی  9789641750338   
3391          Susan Elizabeth Phillips  9780380776832   
4227  Yevgeny Zamyatin, Clarence Brown  9780140185850   
4345                       Dave Eggers  9781932416640   
4567                    Jerzy Kosiński  9780802136343   
4902                  Morris Gleitzman  9780141320632   
5836                   Caroline Kepnes  9781501195433   
6497                          Amy Efaw  9780670011834   

                                                 genres  \
1765  Nonfiction, Autobiography, Memoir, Humor, Audi...   
285   Horror, Fiction, Thriller, Fantasy, Mystery, A...   
905   Fiction, Cultural, Japan, Thriller, Mystery, A...   
1190  Historical, Historical Fiction, Fiction, Audio...   
1243  Romance, Fiction, Contemporary, Womens Fiction...   
1322  Romance, Fiction, Contemporary, Womens Fiction...   
2285  Young Adult, Science Fiction, LGBT, Fiction, D...   
2979  Fiction, Classics, Philosophy, Fantasy, Spirit...   
3391  Romance, Contemporary Romance, Contemporary, S...   
4227  Fiction, Science Fiction, Dystopia, Classics, ...   
4345  Fiction, Cultural, Africa, Historical, Histori...   
4567  Fiction, Humor, Classics, Novels, Literature, ...   
4902  Historical, Historical Fiction, Young Adult, W...   
5836  Thriller, Fiction, Mystery, Mystery Thriller, ...   
6497  Young Adult, Realistic Fiction, Fiction, Conte...   

                                            description  pages  year language  \
1765  From the author of the belovedNew York Timesbe...    240  2015  English   
285   Welcome to Derry, Maine ...It’s a small city, ...   1116  1986  English   
905   Natsuo Kirino's novel tells a story of random ...    400  1997  English   
1190  Memphis, 1939.Twelve-year-old Rill Foss and he...    342  2017  English   
1243  From theNew York Timesbestselling author ofThe...    352  2015  English   
1322  From the #1New York Timesbestselling author of...    369  2012  English   
2285  A boy drowns, desperate and alone in his final...    480  2013  English   
2979                                                NaN    110  1970  English   
3391  What if a woman who knows nothing about sports...    376  1994  English   
4227  The exhilarating dystopian novel that inspired...    226  1924  English   
4345  From the bestselling author ofA Heartbreaking ...    475  2006  English   
4567  A modern classic now available from Grove Pres...    160  1965  English   
4902  Onceby Morris Gleitzman is the story of a youn...    149  2005  English   
5836  When a beautiful aspiring writer strides into ...    422  2014  English   
6497  An infant left in the trash to die. A teenage ...    350  2009  English   

                                            cover_image  total_number_ratings  \
1765  https://i.

**Caution:** The result shows books where the title only contained stopwords, e.g. Stephan King "It". There the title is now emtpy. I need to fill up the ones with empty title as they cannot be vectorized later if they are emtpy.

In [22]:
rows =[6497,5836,4902,4567,4345,4227,3391,2285,1322,1243, 1190,905,285,1765]
for i in rows:
    EnglishBooks['title_tokenized'][i]=EnglishBooks['title'][i]
EnglishBooks[EnglishBooks['title_tokenized']=='']

<ipython-input-22-cb8a8a454392>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  EnglishBooks['title_tokenized'][i]=EnglishBooks['title'][i]


title series                    authors           isbn  \
2979  جاناتان، مرغ دریایی   none  Richard Bach, فاطمه محمدی  9789641750338   

                                                 genres description  pages  \
2979  Fiction, Classics, Philosophy, Fantasy, Spirit...         NaN    110   

      year language                                        cover_image  \
2979  1970  English  https://i.gr-assets.com/images/S/compressed.ph...   

      total_number_ratings  average_rating title_tokenized  \
2979                221751            3.85                   

     desccription_tokenized series_tokenized  
2979                    NaN             none

In [23]:
#check if successful 
EnglishBooks.loc[5836,:]

title                                                                   You
series                                                               You #1
authors                                                     Caroline Kepnes
isbn                                                          9781501195433
genres                    Thriller, Fiction, Mystery, Mystery Thriller, ...
description               When a beautiful aspiring writer strides into ...
pages                                                                   422
year                                                                   2014
language                                                            English
cover_image               https://i.gr-assets.com/images/S/compressed.ph...
total_number_ratings                                                 184107
average_rating                                                         3.92
title_tokenized                                                         You
desccription

#### Remove the last Arabic book at index 2979

In [24]:
EnglishBooks = EnglishBooks.drop(2979)

#### Show final dataframe after title tokenizing

In [25]:
EnglishBooks.shape

(8139, 15)

**After selecting only the English books, there are 8139 of 8431 books left in the database.**

### Preprocess Book Descriptions

In [26]:
EnglishBooksDescr=EnglishBooks.copy()
EnglishBooksDescr.head()

title series  \
283                               Mistress Suffragette   none   
350                         Stones (One True Child #5)   none   
355                                   Mission: Subhero   none   
356  A Summer of Good-Byes (Blue Triangle Press Boo...   none   
357                                  The Scopas Factor   none   

                authors           isbn  \
283      Diana   Forbes     1946409073   
350           L.C. Conn              0   
355  Linda    Armstrong              0   
356      Fred Misurella              0   
357  Vincent Panettiere  9781543947717   

                                                genres  \
283  Fiction, Novels, Drama, Contemporary, Historic...   
350    Book Club, Novels, Fiction, Drama, Contemporary   
355  Contemporary, Novels, Fiction, Drama, Book Clu...   
356  Fiction, Novels, Adult, Contemporary, Adventur...   
357  Fiction, Contemporary, Novels, Drama, Book Clu...   

                                           description  pages  year language  \
283  A young woman without prospects at a ball in G...    333  2017  English   
350  Deep within the New Zealand bush, lies a plate...    327  2019  English   
355  With just a few lines of code, one average man...    225  2017  English   
356  PASSION IN PROVENCE:Ben and Lee Alto follow Va...      0  2017  English   
357  A Hmong "story cloth," a Revolutionary War bat...    326  2018  English   

                                           cover_image  total_number_ratings  \
283  https://i.gr-assets.com/images/S/compressed.ph...                 12973   
350  https://i.gr-assets.com/images/S/compressed.ph...                   569   
355  https://i.gr-assets.com/images/S/compressed.ph...                  2417   
356  https://i.gr-assets.com/images/S/compressed.ph...                  1713   
357  https://i.gr-assets.com/images/S/compressed.ph...                  1080   

     average_rating                              title_tokenized  \
283            4.32                         mistress suffragette   
350            4.31                      stones one true child 5   
355            4.35                              mission subhero   
356            4.36  summer good byes blue triangle press book 2   
357            4.31                                scopas factor   

                                desccription_tokenized series_tokenized  
283  A young woman without prospects at a ball in G...             none  
350  Deep within the New Zealand bush, lies a plate...             none  
355  With just a few lines of code, one average man...             none  
356  PASSION IN PROVENCE:Ben and Lee Alto follow Va...             none  
357  A Hmong "story cloth," a Revolutionary War bat...             none

In [27]:
#Return lower cases
funct.return_Serieslowercase(EnglishBooksDescr,'desccription_tokenized')

In [28]:
#Remove stop words
EnglishBooksDescr['desccription_tokenized']= EnglishBooksDescr.desccription_tokenized.apply(func=funct.remove_stopwords)

In [29]:
#Remove non-Ascii Characters
EnglishBooksDescr['desccription_tokenized'] = EnglishBooksDescr['desccription_tokenized'].apply(func=funct.remove_nonAscii)

In [30]:
#Remove punctuation
EnglishBooksDescr['desccription_tokenized'] = EnglishBooksDescr['desccription_tokenized'].apply(func=funct.remove_punctuation)

In [31]:
EnglishBooksDescr['desccription_tokenized']

283     young woman without prospects ball gilded age ...
350     deep within new zealand bush lies plateau clea...
355     lines code one average man change world and se...
356     passion provence ben lee alto follow van goghs...
357     hmong story cloth revolutionary war battle fla...
                              ...                        
8426    winter 2012 79 year old somerset garden travel...
8427    previously published alternate cover edition a...
8428    100 years journey end vicious cycle true story...
8429    thrive 30 inspirational rags to riches stories...
8430    new york timesbestsellerjeff shaara returns ci...
Name: desccription_tokenized, Length: 8139, dtype: object

#### Check books with no descriptions and insert ''

In [32]:
EnglishBooksDescr[EnglishBooksDescr['desccription_tokenized']=='nan']

title series  \
752                       Last Chance to See   none   
1433                                The Loop   none   
1501  Holy Bible: The New King James Version   none   
1847  Book of Leo: Prequel to Mystical Trash   none   
2187                           A Simple Plan   none   
...                                      ...    ...   
8361                Pictures of Hollis Woods   none   
8369                         The Tree Of Man   none   
8395                    The Last of Her Kind   none   
8398                        South of Reality   none   
8420                    The Elemental Master   none   

                             authors           isbn  \
752   Douglas Adams, Mark Carwardine  9780345371980   
1433                  Nicholas Evans  9780552144957   
1501                       Anonymous  9780718015596   
1847                     Mary Ramsey              0   
2187              Scott        Smith  9780307279958   
...                              ...            ...   
8361            Patricia Reilly Giff  9780440415787   
8369                   Patrick White  9780099324515   
8395          JoAnn Rackear Goldrich  9781505308266   
8398                     A.I. Alexen     9781301529   
8420                    Glen Garrett              0   

                                                 genres description  pages  \
752   Nonfiction, Science, Travel, Environment, Natu...         NaN    256   
1433  Fiction, Romance, Animals, Drama, Contemporary...         NaN    511   
1501  Religion, Christian, Nonfiction, Classics, Ref...         NaN   1536   
1847  Epic, Novels, Suspense, Thriller, Drama, Conte...         NaN     83   
2187  Fiction, Thriller, Mystery, Crime, Suspense, D...         NaN    432   
...                                                 ...         ...    ...   
8361  Young Adult, Realistic Fiction, Fiction, Child...         NaN    166   
8369  Fiction, Cultural, Australia, Novels, Classics...         NaN    480   
8395                                               none         NaN    240   
8398                                               none         NaN      0   
8420                                               none         NaN      0   

      year language                                        cover_image  \
752   1990  English  https://i.gr-assets.com/images/S/compressed.ph...   
1433  1998  English  https://i.gr-assets.com/images/S/compressed.ph...   
1501  1536  English  https://i.gr-assets.com/images/S/compressed.ph...   
1847  2019  English  https://i.gr-assets.com/images/S/compressed.ph...   
2187  1993  English  https://i.gr-assets.com/images/S/compressed.ph...   
...    ...      ...                                                ...   
8361  2002  English  https://i.gr-assets.com/images/S/compressed.ph...   
8369  1955  English  https://i.gr-assets.com/images/S/compressed.ph...   
8395  2015  English  https://i.gr-assets.com/images/S/compressed.ph...   
8398  2012  English  https://i.gr-assets.com/images/S/compressed.ph...   
8420  2013  English  https://i.gr-assets.com/images/S/compressed.ph...   

      total_number_ratings  average_rating                    title_tokenized  \
752                  24202            4.26                    last chance see   
1433                 11567            3.92                               loop   
1501                  7153            4.37  holy bible new king james version   
1847                   568            4.32    book leo prequel mystical trash   
2187                 34654            3.89                        simple plan   
...                    ...             ...                                ...   
8361                 28196            4.02              pictures hollis woods   
8369                  1762            3.95                           tree man   
8395                    22            3.68                          last kind   
8398                     6            4.50                      south reality   


In [33]:
funct.replace_in_column(EnglishBooksDescr, 'desccription_tokenized', 'nan', '')

In [34]:
funct.replace_in_column(EnglishBooksDescr, 'desccription_tokenized', 'none', '')

In [35]:
EnglishBooksDescr[EnglishBooksDescr['desccription_tokenized']=='']

title series  \
752                                  Last Chance to See   none   
935   Understanding China: Learning from China's Pas...   none   
983   Lift: Five Practices Great Managers Do Consist...   none   
1129  The Spirit of Prayer: The Believer's Authority...   none   
1163                           PACKFIRE: Simon Pack # 9   none   
...                                                 ...    ...   
8361                           Pictures of Hollis Woods   none   
8369                                    The Tree Of Man   none   
8395                               The Last of Her Kind   none   
8398                                   South of Reality   none   
8420                               The Elemental Master   none   

                             authors           isbn  \
752   Douglas Adams, Mark Carwardine  9780345371980   
935                     Stefan Piech              0   
983                Raymond   Wheeler              0   
1129                Tolulope Oyewole     1527285502   
1163               John M Vermillion              0   
...                              ...            ...   
8361            Patricia Reilly Giff  9780440415787   
8369                   Patrick White  9780099324515   
8395          JoAnn Rackear Goldrich  9781505308266   
8398                     A.I. Alexen     9781301529   
8420                    Glen Garrett              0   

                                                 genres description  pages  \
752   Nonfiction, Science, Travel, Environment, Natu...         NaN    256   
935   Cultural, China, Politics, Nonfiction, History...        None      0   
983   How To, Nonfiction, Education, Inspirational, ...        None      0   
1129                        Contemporary, Novels, Drama        None      0   
1163               Contemporary, Fiction, Drama, Novels        None      0   
...                                                 ...         ...    ...   
8361  Young Adult, Realistic Fiction, Fiction, Child...         NaN    166   
8369  Fiction, Cultural, Australia, Novels, Classics...         NaN    480   
8395                                               none         NaN    240   
8398                                               none         NaN      0   
8420                                               none         NaN      0   

      year language                                        cover_image  \
752   1990  English  https://i.gr-assets.com/images/S/compressed.ph...   
935   2020  English  https://i.gr-assets.com/images/S/compressed.ph...   
983   2020  English  https://i.gr-assets.com/images/S/compressed.ph...   
1129  2021  English  https://i.gr-assets.com/images/S/compressed.ph...   
1163  2020  English  https://i.gr-assets.com/images/S/compressed.ph...   
...    ...      ...                                                ...   
8361  2002  English  https://i.gr-assets.com/images/S/compressed.ph...   
8369  1955  English  https://i.gr-assets.com/images/S/compressed.ph...   
8395  2015  English  https://i.gr-assets.com/images/S/compressed.ph...   
8398  2012  English  https://i.gr-assets.com/images/S/compressed.ph...   
8420  2013  English  https://i.gr-assets.com/images/S/compressed.ph...   

      total_number_ratings  average_rating  \
752                  24202            4.26   
935                     32            4.13   
983                    437            4.38   
1129                   553            4.30   
1163                  1126            4.31   
...                    ...             ...   
8361                 28196            4.02   
8369                  1762            3.95   
8395                    22            3.68   
8398                     6            4.50   
8420                     6            4.83   

                                        title_tokenized  \
752                                     last chance see   
935   understanding china learning china past presen...   
983   lift five practices great managers

### Preprocess Column Series

In [36]:
EnglishBooksSeries= EnglishBooksDescr.copy()

In [37]:
#Return lower cases
funct.return_Serieslowercase(EnglishBooksSeries,'series_tokenized')

In [38]:
#Remove stop words
EnglishBooksSeries['series_tokenized']= EnglishBooksSeries.series_tokenized.apply(func=funct.remove_stopwords)

In [39]:
#Remove non-Ascii Characters
EnglishBooksSeries['series_tokenized'] = EnglishBooksSeries['series_tokenized'].apply(func=funct.remove_nonAscii)

In [40]:
#Remove punctuation
EnglishBooksSeries['series_tokenized'] = EnglishBooksSeries['series_tokenized'].apply(func=funct.remove_punctuation)

In [41]:
EnglishBooksSeries['series_tokenized']

283                                      none
350                                      none
355                                      none
356                                      none
357                                      none
                        ...                  
8426                                     none
8427                     treemakers trilogy 1
8428                                     none
8429                                     none
8430    civil war 1861 1865 western theater 3
Name: series_tokenized, Length: 8139, dtype: object

#### Check all books with no series and enter ''

In [42]:
EnglishBooksSeries[EnglishBooksSeries['series_tokenized']=='none']

title series  \
283                                Mistress Suffragette   none   
350                          Stones (One True Child #5)   none   
355                                    Mission: Subhero   none   
356   A Summer of Good-Byes (Blue Triangle Press Boo...   none   
357                                   The Scopas Factor   none   
...                                                 ...    ...   
8423                     The Testament of Vida Tremayne   none   
8425                                   Roommate of Fire   none   
8426                                   The Stove-Junker   none   
8428        100 Years: A Journey to End a Vicious Cycle   none   
8429    Thrive: 30 Inspirational Rags-to-Riches Stories   none   

                 authors           isbn  \
283       Diana   Forbes     1946409073   
350            L.C. Conn              0   
355   Linda    Armstrong              0   
356       Fred Misurella              0   
357   Vincent Panettiere  9781543947717   
...                  ...            ...   
8423       Sarah Vincent  9781910153109   
8425         Aiden Bates              0   
8426          S.K. Kalsi  9780990779063   
8428     Mark L. Baynard  9780986138003   
8429       Jason Navallo     1511503882   

                                                 genres  \
283   Fiction, Novels, Drama, Contemporary, Historic...   
350     Book Club, Novels, Fiction, Drama, Contemporary   
355   Contemporary, Novels, Fiction, Drama, Book Clu...   
356   Fiction, Novels, Adult, Contemporary, Adventur...   
357   Fiction, Contemporary, Novels, Drama, Book Clu...   
...                                                 ...   
8423                                    Fiction, Novels   
8425                               Romance, M M Romance   
8426                                               none   
8428                                               none   
8429  Nonfiction, Self Help, Inspirational, Business...   

                                            description  pages  year language  \
283   A young woman without prospects at a ball in G...    333  2017  English   
350   Deep within the New Zealand bush, lies a plate...    327  2019  English   
355   With just a few lines of code, one average man...    225  2017  English   
356   PASSION IN PROVENCE:Ben and Lee Alto follow Va...      0  2017  English   
357   A Hmong "story cloth," a Revolutionary War bat...    326  2018  English   
...                                                 ...    ...   ...      ...   
8423  A lonely novelist, A devoted fan, A journal th...    400  2014  English   
8425  A Gay Romance Erotica MMNeil, a man desperate ...     32  2015  English   
8426  In the winter of 2012, 79-year-old Somerset Ga...    338  2015  English   
8428  100 Years, A Journey to End a Vicious Cycle is...    290  2015  English   
8429  Thrive: 30 Inspirational Rags-to-Riches Storie...    100  2015  English   

                                            cover_image  total_number_ratings  \
283   https://i.gr-assets.com/images/S/compressed.ph...                 12973   
350   https://i.gr-assets.com/images/S/compressed.ph...                   569   
355   https://i.gr-assets.com/images/S/compressed.ph...                  2417   
356   https://i.gr-assets.com/images/S/compressed.ph...                  1713   
357   https://i.gr-assets.com/images/S/compressed.ph...                  1080   
...                                                 ...                   ...   
8423  https://i.gr-assets.com/images/S/compressed.ph...                    93   
8425  https://i.gr-assets.com/images/S/compressed.ph...                    43   
8426  https://i.gr-assets.com/images/S/compressed.ph...                    23   
8428  https://i.gr-assets.com/images/S/compressed.ph...                    13   
8429  https://i.gr-assets.com/images/S/compressed.ph...                   335   

      average_rating                                 title_tokenized  \
283          

In [43]:
EnglishBooksSeries[EnglishBooksSeries['series_tokenized']=='']

title            series  \
5233  Stories Of The Prophets  البداية والنهاية   
6794     The End of the World      نهاية العالم   

                                       authors           isbn  \
5233  ابن كثير, Ibn Kathir, Rashad Ahmad Azami  9789960892269   
6794                    محمد عبدالرحمن العريفي              0   

                                                 genres  \
5233  Religion, Islam, History, Nonfiction, Biograph...   
6794                        Religion, Islam, Nonfiction   

                                            description  pages  year language  \
5233  The Stories of the Prophets have been compiled...    576  1981  English   
6794  In recent times things have become very confus...    410  2010  English   

                                            cover_image  total_number_ratings  \
5233  https://i.gr-assets.com/images/S/compressed.ph...                  5749   
6794  https://i.gr-assets.com/images/S/compressed.ph...                  4105   

      average_rating   title_tokenized  \
5233            4.49  stories prophets   
6794            4.13         end world   

                                 desccription_tokenized series_tokenized  
5233  stories prophets compiled al bidayah wan nihay...                   
6794  recent times things become confusing begun see...

#### Delete another 2 found arabic books at index 5233 and 6794

In [44]:
in_drop=[5233,6794]
EnglishBooksSeries = EnglishBooksSeries.drop(in_drop)
EnglishBooksSeries.shape

(8137, 15)

**After removing these two books, there are 8137 books left in the dataset.**

#### Replace all books with no series with ''

In [45]:
funct.replace_in_column(EnglishBooksSeries, 'series_tokenized', 'none', '')
EnglishBooksSeries['series_tokenized'].value_counts()

                               5398
1                                10
chronicles narnia                 6
james bond                        5
foundation                        5
                               ... 
villains 1                        1
graceling realm 2                 1
criminal profiler 3               1
selection 1                       1
celtic cousins adventures 2       1
Name: series_tokenized, Length: 2616, dtype: int64

In [46]:
EnglishBooksSeries[EnglishBooksSeries['series'].str.match('Harry Potter')]

title           series  \
1        Harry Potter and the Sorcerer's Stone  Harry Potter #1   
13        Harry Potter and the Deathly Hallows  Harry Potter #7   
25    Harry Potter and the Prisoner of Azkaban  Harry Potter #3   
30         Harry Potter and the Goblet of Fire  Harry Potter #4   
33      Harry Potter and the Half-Blood Prince  Harry Potter #6   
73     Harry Potter and the Chamber of Secrets  Harry Potter #2   
279  Harry Potter and the Order of the Phoenix  Harry Potter #5   

                                       authors           isbn  \
1                                 J.K. Rowling              0   
13                                J.K. Rowling      747591059   
25                 J.K. Rowling, Mary GrandPré  9780439655484   
30                 J.K. Rowling, Mary GrandPré              0   
33                                J.K. Rowling      747581088   
73   J.K. Rowling, Mary GrandPré, ویدا اسلامیه     1408855666   
279                J.K. Rowling, Mary GrandPré      747551006   

                                                genres  \
1    Fantasy, Fiction, Young Adult, Magic, Children...   
13   Fantasy, Young Adult, Fiction, Magic, Children...   
25   Fantasy, Young Adult, Fiction, Magic, Children...   
30   Fantasy, Young Adult, Fiction, Magic, Children...   
33   Fantasy, Young Adult, Fiction, Magic, Children...   
73   Fantasy, Young Adult, Fiction, Magic, Children...   
279  Fantasy, Young Adult, Fiction, Magic, Children...   

                                           description  pages  year language  \
1    Harry Potter's life is miserable. His parents ...    309  1997  English   
13   It's no longer safe for Harry at Hogwarts, so ...    759  2007  English   
25   For twelve long years, the dread fortress of A...    435  1999  English   
30   Harry Potter is midway through his training as...    734  2000  English   
33   The war against Voldemort is not going well; e...    652  2005  English   
73   Ever since Harry Potter had come home for the ...    341  1998  English   
279  There is a door at the end of a silent corrido...    870  2003  English   

                                           cover_image  total_number_ratings  \
1    https://i.gr-assets.com/images/S/compressed.ph...               8003482   
13   https://i.gr-assets.com/images/S/compressed.ph...               3111326   
25   https://i.gr-assets.com/images/S/compressed.ph...               3223918   
30   https://i.gr-assets.com/images/S/compressed.ph...               2907124   
33   https://i.gr-assets.com/images/S/compressed.ph...               2714400   
73   https://i.gr-assets.com/images/S/compressed.ph...               3089063   
279  https://i.gr-assets.com/images/S/compressed.ph...               2800355   

     average_rating                 title_tokenized  \
1              4.47   harry potter sorcerer s stone   
13             4.61    harry potter deathly hallows   
25             4.57   harry potter prisoner azkaban   
30             4.56        harry potter goblet fire   
33             4.57  harry potter half blood prince   
73             4.43    harry potter chamber secrets   
279            4.50      harry potter order phoenix   

                                desccription_tokenized series_tokenized  
1    harry potter s life miserable parents dead he ...   harry potter 1  
13   longer safe harry hogwarts best friends ron he...   harry potter 7  
25   twelve long years dread fortress azkaban held ...   harry potter 3  
30   harry potter midway training wizard coming age...   harry potter 4  
33   war voldemort going well even muggle governmen...   harry potter 6  
73   ever since harry potter come home summer dursl...   harry potter 2  
279  door end silent corridor its haunting harry po...   harry potter 5

## Save to Csv file

In [47]:
EnglishBooksSeries.to_csv('DataFinal\DataPreprocessed.csv')